crates/vm/src/expr.rs

```diff
@@ -315,19 +490,238 @@ impl QueryExpr {
         }
     }
 
-    pub fn with_index_scan(self, col_id: u32, value: AlgebraicValue, table: DbTable) -> Self {
-        let mut expr = self;
-        expr.query.push(Query::IndexScan(col_id, value, table));
-        expr
+    // Generate an index scan for an equality predicate if this is the first
+    // operator. Otherwise generate a select.
+    pub fn with_index_eq(mut self, table: DbTable, col_id: u32, value: AlgebraicValue) -> Self {
```

Comment:
If is too complicated to generalize to all predicates then I think we need a TODO here. We need to collect all the issues we found so in the refactor we solve them.